### ISSUE TITLE
BoundSharedState changed events is misnamed (or misleading)

### ISSUE THREAD

**Issue by addyess** [canonical/operator#225]
Here's a simple object

```python
class ExampleConfig(Object):
    state = StoredState()

    def __init__(self, *args):
        super().__init__(*args)
        self.state.set_default(conf=dict())
        self.framework.observe(self.state.on.changed, self)

    def on_changed(self, event):
        event.log("When does this fire?")
```
I expected changing the value of `state.conf['item'] = 1` would have triggered a change event... but no dice. It seems the on.changed event is only emited on `BoundStoredState.__setattr__`

Really this 'changed' should be called 'assigned'

```python
class ExampleConfig(Object):
    state = StoredState()

    def __init__(self, *args):
        super().__init__(*args)
        self.state.set_default(conf=dict())
        self.framework.observe(self.state.on.assigned, self)

    def on_assigned(self, event):
        # which attribute was assigned??  its not in the event
        # what was it changed from/to?  it's also not in the event
        # how is this useful?  Is there good documentation on how to use it?
        event.log("at least i know SOMETHING was assigned")
```

**Comment by jameinel** [comment]
I think it would be possible to trigger changed to the attributes of a stored dict. `__getattr__` is already returning a wrapped object so that we can track the `dirty` flag. It is just a matter of wherever we set dirty to also trigger changed. It is a little bit trickier to give you the concrete 'what changed' as you can end up with nesting/sets/lists/dicts, etc. If you did:
```
myattr = {'a': [{}, {}, {}, set('b')}
self.state.setdefault(foo=myattr)
```

And then later do
```
self.state.foo['a'][3].add('c')
```
What would be a 'reasonable' way of telling you what has changed. The state certainly has changed, but would 'foo.a.3' reasonably convey that a member of your set has changed?

I could very much see conveying the top level state attribute that hash changed. So if you have:

```
self.state.set_default(foo=dict(), bar=set(), baz=2)
```

And then later
```
self.state.foo['a'] = 2
```
the event could contain
```
event.attribute = 'foo'
```

However, if you did
```
self.state.foo['a'] = {}
self.state.foo['a']['b'] = 2
```
You would still just see `foo` as the attribute that has changed.

I think that would cover your basic use cases, and avoid overly complicating things. If you want fine-grained notification, then you make them top level attributes instead of deeply nested.

Another thing to consider, should set_default trigger a change event if it actually changes data. (imagine a charm upgrade, etc.) For now, I'm going to stick with set_default doesn't trigger changes, but it is something we might want to revisit.

**Comment by jameinel** [comment]
so pull #226 potentially addresses this (certainly it would fix nested updates, etc).
But it would be good to understand your actual use case. In general, we've tried to push having StoredState be private to the type itself, and not something that you would use externally to 'snoop' on what is going on.

If you want to know when something happens, the recommendation would be to have an explicit named event for it, rather than a side effect of 'something in state has changed'. That also allows you to potentially rename the stored state attribute without breaking anyone peeking at what has been changing in your state.

**Comment by jameinel** [comment]
PR #227 'fixes' this by removing the incorrect event altogether

**Comment by niemeyer** [comment]
@addyess I think it's both misnamed and misleading :-)

This attribute is a left over from our original implementation that would use events to snapshot the data into the database. Back then it would indeed trigger the event in all cases, including inner mutable values. The problem with that approach was that it was unnecessarily heavy.. every small touch of the "_stored" attribute would cause a database roundtrip. So we changed that and made it only persist into the database once right before the commit takes place, and only if the value was changed.

We should probably just get rid of that attribute by now, since it's indeed misleading and I would even argue that it doesn't lead to good code, because the event is not enough to take a position on what indeed is changing. It's much better for the component to raise a higher-level event specific to the business logic.  

My vote is to kill the event for now.
